File: my-solutions/0208-implement-trie-prefix-tree/solution.cpp
class Trie {
    struct Node {
        Node* c[26] = {nullptr}; // Fixed-size array for children
        bool e = false;          // End-of-word flag
    } root;  
public:
    Trie() {
     
    }
    
    void insert(string w) {
        Node* n = &root;
        for (char ch : w) {
            if (!n->c[ch - 'a']) n->c[ch - 'a'] = new Node();
            n = n->c[ch - 'a'];
        }
        n->e = true;
    }
    
    bool search(string w) const {
        const Node* n = &root;
        for (char ch : w) {
            if (!n->c[ch - 'a']) return false;
            n = n->c[ch - 'a'];
        }
        return n->e;
    }
    
    bool startsWith(string p) const {
        const Node* n = &root;
        for (char ch : p) {
            if (!n->c[ch - 'a']) return false;
            n = n->c[ch - 'a'];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: my-solutions/0208-implement-trie-prefix-tree/solution.py
class Node:
    def __init__(self, end = 'False'):
        self.end = end
        self.next = [None] * 26



class Trie:

    def __init__(self):
        self.root = Node()
        

    def insert(self, word: str) -> None:
        curr = self.root
        for char in word:
            if not curr.next[ord(char) - ord('a')]:
                curr.next[ord(char) - ord('a')] = Node()
            curr = curr.next[ord(char) - ord('a')]
        curr.end = True
        

    def search(self, word: str) -> bool:
        curr = self.root
        for char in word:
            if not curr.next[ord(char) - ord('a')]:
                print("")
                return False
            curr = curr.next[ord(char) - ord('a')]        
        return curr.end == True
        

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for char in prefix:
            if not curr.next[ord(char) - ord('a')]:
                return False
            curr = curr.next[ord(char) - ord('a')]
        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
